### PR TITLE
Issue #3295 - Add R4B to $versions operation

### PR DESCRIPTION
**Pull request by tbieste**
Issue #3295 - Add R4B to $versions operation

Signed-off-by: Troy Biesterfeld <tbieste@us.ibm.com>

### CODE REVIEW

File: operation/fhir-operation-validate/pom.xml [hunk 1]
```diff
@@ -19,6 +19,12 @@
             <version>${project.version}</version>
             <scope>provided</scope>
         </dependency>
+        <dependency>
+            <groupId>${project.groupId}</groupId>
+            <artifactId>fhir-server-test</artifactId>
+            <version>${project.version}</version>
+            <scope>test</scope>
+        </dependency>
```
Comment (prb112) [inline operation/fhir-operation-validate/pom.xml:27]: Any reason for this update?

Comment (prb112) [reply operation/fhir-operation-validate/pom.xml:27]: Seems ancillary to the intent of the PR

File: fhir-server-test/src/test/java/com/ibm/fhir/server/test/operation/VersionsOperationTest.java [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+import com.ibm.fhir.server.test.FHIRServerTestBase;
+
+import jakarta.json.JsonArray;
+import jakarta.json.JsonObject;
+
+/**
+ * This class tests the $versions operation, which returns the list of supported FHIR versions,
+ * along with the default FHIR version.
+ */
+public class VersionsOperationTest extends FHIRServerTestBase {
```
Comment (prb112) [inline fhir-server-test/src/test/java/com/ibm/fhir/server/test/operation/VersionsOperationTest.java:39]: Do you see this as being useful for other tests? that is knowing the version the server supports? 

if so, maybe parts of this should be pushed down to FHIRServerTestBase

File: operation/fhir-operation-versions/src/main/java/com/ibm/fhir/operation/versions/VersionsOperation.java [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+
+            // Build and return response based on the Accept header
+            String outputFormat = getNonFhirOutputFormat(operationContext);
+            if (MediaType.APPLICATION_JSON.equals(outputFormat)) {
+                return buildJsonResponse(operationContext, versions, defaultVersion);
+            } else if (MediaType.APPLICATION_XML.equals(outputFormat)) {
+                return buildXmlResponse(operationContext, versions, defaultVersion);
+            } else {
+                return buildFhirResponse(versions, defaultVersion);
+            }
```
Comment (prb112) [inline operation/fhir-operation-versions/src/main/java/com/ibm/fhir/operation/versions/VersionsOperation.java:84]: Why not defer to FHIRProvider, it'll transpose from the model to the output format for the version based on the Accept Header? Then the framework will change it. 

Is this due to a non-standard output?

File: operation/fhir-operation-validate/pom.xml (hunk 1 above)
Comment (tbieste) [reply operation/fhir-operation-validate/pom.xml]: Cleaned up during rebase.

File: operation/fhir-operation-versions/src/main/resources/META-INF/services/com.ibm.fhir.server.spi.operation.FHIROperation [hunk 4]
```diff
@@ -0,0 +1 @@
+com.ibm.fhir.operation.versions.VersionsOperation
```
[the remark this answers is not in the document]
Comment (tbieste) [reply operation/fhir-operation-versions/src/main/resources/META-INF/services/com.ibm.fhir.server.spi.operation.FHIROperation]: Cleaned up.

File: fhir-server-test/src/test/java/com/ibm/fhir/server/test/operation/VersionsOperationTest.java (hunk 2 above)
Comment (tbieste) [reply fhir-server-test/src/test/java/com/ibm/fhir/server/test/operation/VersionsOperationTest.java]: Maybe.  Not sure at this point.  I'd say if we see a need for that in the future, to move/add common logic at that point.

File: operation/fhir-operation-versions/src/main/java/com/ibm/fhir/operation/versions/VersionsOperation.java (hunk 3 above)
Comment (tbieste) [reply operation/fhir-operation-versions/src/main/java/com/ibm/fhir/operation/versions/VersionsOperation.java]: I added a README.md to explain the various response formats.

File: operation/fhir-operation-versions/src/main/java/com/ibm/fhir/operation/versions/VersionsOperation.java [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+import jakarta.json.JsonObject;
+
+/**
+ * Implementation of the $versions operation, which returns the list of supported FHIR versions,
+ * along with the default FHIR version.
+ */
+public class VersionsOperation extends AbstractOperation {
+
+    private static final JsonBuilderFactory JSON_BUILDER_FACTORY = Json.createBuilderFactory(null);
+    private static final DocumentBuilderFactory XML_DOM_FACTORY = DocumentBuilderFactory.newInstance();
```
[the remark this answers is not in the document]
Comment (tbieste) [reply operation/fhir-operation-versions/src/main/java/com/ibm/fhir/operation/versions/VersionsOperation.java]: Updated.

File: operation/fhir-operation-versions/src/main/java/com/ibm/fhir/operation/versions/VersionsOperation.java [hunk 6]
```diff
@@ -82,6 +91,23 @@ protected Parameters doInvoke(FHIROperationContext operationContext, Class<? ext
...
+    private List<String> getFhirVersions() {
+        return Stream.of(FHIRVersionParam.values()).map(k -> k.value()).collect(Collectors.toList());
+    }
+
+    /**
+     * Gets the default FHIR version based on the server configuration.
+     * @return the default FHIR version
+     */
+    private String getDefaultFhirVersion() {
+        String fhirVersionString = FHIRConfigHelper.getStringProperty(FHIRConfiguration.PROPERTY_DEFAULT_FHIR_VERSION, VERSION_40.value());
```
Comment (lmsurpre) [inline operation/fhir-operation-versions/src/main/java/com/ibm/fhir/operation/versions/VersionsOperation.java:107]: perfect

File: operation/fhir-operation-versions/README.md [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+
+```xml
+HTTP/1.1 200 OK
+[other headers]
+
+<Parameters xmlns="http://hl7.org/fhir">
+  <parameter>
+		<name value="version"/>
+		<valueString value="4.0"/>
+  </parameter>
```
Comment (lmsurpre) [inline operation/fhir-operation-versions/README.md:60]: funky mix of tabs and spaces here...i'd suggest updating to use spaces even though its just doc

File: operation/fhir-operation-versions/README.md [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+
+```json
+HTTP/1.1 200 OK
+[other headers]
+
+{
+    "resourceType": "Parameters",
+    "parameter": [{
+    "name": "version",
+    "valueString": "4.0"
```
Comment (lmsurpre) [inline operation/fhir-operation-versions/README.md:29]: maybe indent with 4 spaces so the structure stands out better and matches our actually pretty-printed response

**Review by lmsurpre — APPROVED** [review]

File: operation/fhir-operation-versions/README.md (hunk 7 above)
Comment (tbieste) [reply operation/fhir-operation-versions/README.md]: Fixed.

File: operation/fhir-operation-versions/README.md (hunk 8 above)
Comment (tbieste) [reply operation/fhir-operation-versions/README.md]: Fixed.

**Pull request merged**